File: src/tablet/otd/parsers/xpPen.ts
/** XP-Pen vendor-specific report parsers. */

import {
  OutOfRangeReport,
  TabletReport,
  Tuple2,
  bit,
  s8,
  u16le,
} from "../reports.js";
import { ParsedReport, ReportParser } from "./generic.js";

export class XPPenTabletReport {
  readonly kind = "XPPenTabletReport" as const;
  raw: Uint8Array;
  position: Tuple2 = [0, 0];
  pressure = 0;
  tilt: Tuple2 = [0, 0];
  penButtons: boolean[] = [];
  eraser = false;

  constructor(raw: Uint8Array) { this.raw = raw; }

  static parse(data: Uint8Array): XPPenTabletReport {
    const r = new XPPenTabletReport(data);
    r.position = [u16le(data, 2), u16le(data, 4)];
    r.pressure = u16le(data, 6);
    r.tilt = [s8(data[8]), s8(data[9])];
    r.penButtons = [bit(data[1], 1), bit(data[1], 2)];
    r.eraser = bit(data[1], 3);
    return r;
  }
}

/**
 * 24-bit X/Y position assembled from low ushort + high byte from tilt slots.
 */
export class XPPenTabletOverflowReport {
  readonly kind = "XPPenTabletOverflowReport" as const;
  raw: Uint8Array;
  position: Tuple2 = [0, 0];
  pressure = 0;
  tilt: Tuple2 = [0, 0];
  penButtons: boolean[] = [];
  eraser = false;

  constructor(raw: Uint8Array) { this.raw = raw; }

  static parse(data: Uint8Array): XPPenTabletOverflowReport {
    const r = new XPPenTabletOverflowReport(data);
    const x = u16le(data, 2) | (data[10] << 16);
    const y = u16le(data, 4) | (data[11] << 16);
    r.position = [x, y];
    r.pressure = u16le(data, 6);
    r.tilt = [s8(data[8]), s8(data[9])];
    r.penButtons = [bit(data[1], 1), bit(data[1], 2)];
    r.eraser = bit(data[1], 3);
    return r;
  }
}

export class XPPenAuxReport {
  readonly kind = "XPPenAuxReport" as const;
  raw: Uint8Array;
  auxButtons: boolean[] = [];
  analogDeltas: number[] = [];

  constructor(raw: Uint8Array) { this.raw = raw; }

  static parse(data: Uint8Array, auxIndex = 2, wheelIndex = 7): XPPenAuxReport {
    const r = new XPPenAuxReport(data);
    const buttons: boolean[] = [];
    for (let i = 0; i < 8; i++) buttons.push(bit(data[auxIndex], i));
    for (let i = 0; i < 8; i++) buttons.push(bit(data[auxIndex + 1], i));
    for (let i = 0; i < 4; i++) buttons.push(bit(data[auxIndex + 2], i));
    r.auxButtons = buttons;
    const w = data[wheelIndex];
    r.analogDeltas = [
      bit(w, 0) ? 1 : bit(w, 1) ? -1 : 0,
      bit(w, 4) ? 1 : bit(w, 5) ? -1 : 0,
    ];
    return r;
  }
}

export class XPPenReportParser implements ReportParser {
  parse(data: Uint8Array): ParsedReport {
    if (data[1] === 0xC0) return new OutOfRangeReport(data);
    if (bit(data[1], 4)) return XPPenAuxReport.parse(data);
    if (data.length >= 12) return XPPenTabletOverflowReport.parse(data);
    if (data.length >= 10) return XPPenTabletReport.parse(data);
    return TabletReport.parse(data);
  }
}

File: python/sketchatone/tablet/otd/parsers/xp_pen.py
"""XP-Pen vendor-specific report parsers."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Tuple

from ..reports import OutOfRangeReport, TabletReport, _bit, _s8, _u16le


@dataclass
class XPPenTabletReport:
    """Port of OpenTabletDriver.Configurations.Parsers.XP_Pen.XP_PenTabletReport."""
    raw: bytes
    position: Tuple[int, int] = (0, 0)
    pressure: int = 0
    tilt: Tuple[int, int] = (0, 0)
    pen_buttons: List[bool] = field(default_factory=list)
    eraser: bool = False

    @classmethod
    def parse(cls, data: bytes) -> "XPPenTabletReport":
        return cls(
            raw=data,
            position=(_u16le(data, 2), _u16le(data, 4)),
            pressure=_u16le(data, 6),
            tilt=(_s8(data[8]), _s8(data[9])),
            pen_buttons=[_bit(data[1], 1), _bit(data[1], 2)],
            eraser=_bit(data[1], 3),
        )


@dataclass
class XPPenTabletOverflowReport:
    """Port of OpenTabletDriver.Configurations.Parsers.XP_Pen.XP_PenTabletOverflowReport.

    24-bit X/Y position assembled from low ushort + high byte from tilt slots.
    """
    raw: bytes
    position: Tuple[int, int] = (0, 0)
    pressure: int = 0
    tilt: Tuple[int, int] = (0, 0)
    pen_buttons: List[bool] = field(default_factory=list)
    eraser: bool = False

    @classmethod
    def parse(cls, data: bytes) -> "XPPenTabletOverflowReport":
        x = _u16le(data, 2) | (data[10] << 16)
        y = _u16le(data, 4) | (data[11] << 16)
        return cls(
            raw=data,
            position=(x, y),
            pressure=_u16le(data, 6),
            tilt=(_s8(data[8]), _s8(data[9])),
            pen_buttons=[_bit(data[1], 1), _bit(data[1], 2)],
            eraser=_bit(data[1], 3),
        )


@dataclass
class XPPenAuxReport:
    """Port of OpenTabletDriver.Configurations.Parsers.XP_Pen.XP_PenAuxReport."""
    raw: bytes
    aux_buttons: List[bool] = field(default_factory=list)
    analog_deltas: List[int] = field(default_factory=list)

    @classmethod
    def parse(cls, data: bytes, aux_index: int = 2, wheel_index: int = 7) -> "XPPenAuxReport":
        buttons = (
            [_bit(data[aux_index], i) for i in range(8)]
            + [_bit(data[aux_index + 1], i) for i in range(8)]
            + [_bit(data[aux_index + 2], i) for i in range(4)]
        )
        w = data[wheel_index]
        deltas = [
            1 if _bit(w, 0) else (-1 if _bit(w, 1) else 0),
            1 if _bit(w, 4) else (-1 if _bit(w, 5) else 0),
        ]
        return cls(raw=data, aux_buttons=buttons, analog_deltas=deltas)


class XPPenReportParser:
    """Port of OpenTabletDriver.Configurations.Parsers.XP_Pen.XP_PenReportParser."""

    def parse(self, data: bytes):
        if data[1] == 0xC0:
            return OutOfRangeReport(raw=data)
        if _bit(data[1], 4):
            return XPPenAuxReport.parse(data)
        if len(data) >= 12:
            return XPPenTabletOverflowReport.parse(data)
        if len(data) >= 10:
            return XPPenTabletReport.parse(data)
        return TabletReport.parse(data)
